pkg/metrics/controller-manager.go

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/client-go/util/workqueue"
+
+	gardeninformers "github.com/gardener/gardener/pkg/client/garden/informers/externalversions/garden/v1beta1"
+	kubeinformers "k8s.io/client-go/informers/core/v1"
+)
+
+var (
+	controllerManagerMetricsBucket      = make([]prometheus.Metric, 0)
+	controllerManagerMetricsBucketMutex sync.Mutex
+
+	// ControllerManagerWorkerSumDesc is a prometheus desc for the worker sum of a controller.
+	ControllerManagerWorkerSumDesc = prometheus.NewDesc("garden_cm_worker_total", "Count of running controller workers", []string{"controller"}, nil)
+
+	// FlowDuration is a desc for Prometheus metrics for the duration of a flow in seconds.
+	FlowDuration = prometheus.NewDesc("garden_cm_flow_duration_seconds", "Duration of a flow execution.", []string{"name", "resource"}, nil)
+
+	// FlowResult is a desc Prometheus metrics for result of flow execution.
+	FlowResult = prometheus.NewDesc("garden_cm_flow_results", "Results of a flow excecution.", []string{"name", "resource"}, nil)
+)
+
+// InitControllerManagerMetrics takes a garden kubernetes client to register Garden
+// and Garden Controller Manager related metrics.
+func InitControllerManagerMetrics(shootInformer gardeninformers.ShootInformer, namespaceInformer kubeinformers.NamespaceInformer) {
```

Comment:
use pointers and don't so `shootInformer` and `namespaceInformer` is not copied when this method is called